src/gc-arena/src/context.rs

```diff
@@ -324,6 +377,57 @@ impl Context {
             }
         }
     }
+
+    /// Determines whether or not a Gc pointer is safe to be upgraded.
+    /// This is used by weak pointers to determine if it can safely upgrade to a strong pointer.
+    ///
+    /// Safety: `ptr` must be a valid pointer to a GcBox<T>.
+    unsafe fn upgrade<T: Collect>(&self, ptr: NonNull<GcBox<T>>) -> bool {
+        let gc_box = ptr.as_ref();
+
+        // This object has already been freed, definitely not safe to upgrade.
+        if !gc_box.flags.alive() {
+            return false;
+        }
+
+        // Consider the different possible phases of the GC:
+        // * In `Phase:Wake` or `Phase::Sleep`, the GC is not running, so we can upgrade.
+        //   If the newly-created `Gc` or `GcCell` survives the current `arena.mutate`
+        //   call, then the situtation is equivalent to having copied an existing `Gc`/`GcCell`,
+        //   or having created a new allocation.
+        //
+        // * In `Phase::Propagate`:
+        //   If the newly-created `GcWeak` or `GcCell` survives the current `arena.mutate`
+        //   call, then it must have been stored somewhere, triggering a write barrier.
+        //   This will ensure that the new `Gc`/`GcCell` gets traced (if it's now reachable)
+        //   before we transition to `Phase::Sweep`.
+        //
+        // * In `Phase::Sweep`:
+        //   If the allocation is white and has the `has_weak_ref()` flag set, then it's impossile
+        //   for it to have been freshly-created during this `Phase::Sweep`. `has_weak_ref_` is only
+        //   set when a `GcWeak/GcWeakCell` is traced. A `GcWeak/GcWeakCell` must be created from
+        //   an existing `Gc/GcCell` via `downgrade()`, so `has_weak_ref() == true` means that a
+        //   `GcWeak/GcWeakCell` existing during the last `Phase::Sweep.`
```

Comment:
The overall logic is not wrong, but this isn't precisely true, right? `has_weak_ref() == true` means that a weak pointer existed at some point during the previous `Wake` or `Propagate` phases, right? It could have been created then?

nit: also.. `GcWeak/GcWeakCell` -> `GcWeak` / `GcWeakCell` but it's not a doc comment so this is super just a nit.

Reply:
You're right - I don't know why I wrote that. I've updated the comment.